pytest_operator/plugin.py

```diff
@@ -347,53 +364,184 @@ def download_resource(self, resource, destination: Path):
+
+
 class OpsTest:
     """Utility class for testing Operator Charms."""
 
-    _instance = None  # store instance so we can tell if it's been used yet
+    _instance = None  # store instance, so we can tell if it's been used yet
 
     def __init__(self, request, tmp_path_factory):
         self.request = request
-        self.tmp_path: Path = tmp_path_factory.mktemp(self.default_model_name)
-        log.info(f"Using tmp_path: {self.tmp_path}")
+        self._tmp_path_factory = tmp_path_factory
+        self._global_tmp_path = None
 
         # Flag indicating whether all subsequent tests should be aborted.
         self.aborted = False
 
         # Flag for using destructive mode or not for charm builds.
         self.destructive_mode = request.config.option.destructive_mode
 
+        # Config options to determine first model specs used by tests
+        self._orig_model_alias = request.config.option.model_alias
+        self._init_cloud_name = request.config.option.cloud
+        self._init_model_name = request.config.option.model
+        self._init_keep_model = request.config.option.keep_models
+
         # These may be modified by _setup_model
-        self.cloud_name = request.config.option.cloud
         self.controller_name = request.config.option.controller
-        self.model_name = request.config.option.model
-        self.keep_model = request.config.option.keep_models
-        self.model_config = request.config.option.model_config
+        self._init_model_config = request.config.option.model_config
 
         # Flag for enabling the juju-crashdump
         self.crash_dump = not request.config.option.no_crash_dump
         self.crash_dump_output = request.config.option.crash_dump_output
 
         # These will be set by _setup_model
-        self.model_full_name = None
-        self.model = None
         self.jujudata = None
-        self._controller = None
+        self._controller: Optional[Controller] = None
+
+        # maintains a set of all models connected by this fixture
+        # use an OrderedDict so that the first model made
+        self._current_alias = None
+        self._models: MutableMapping[str, ModelState] = OrderedDict()
+
+    @contextlib.contextmanager
+    def model_context(self, alias: str) -> Generator[Model, None, None]:
+        """
+        Analog to `juju switch` where the focus of the current model is moved.
+        """
+        prior = self.current_alias
+        model = self._switch(alias)
+        try:
+            yield model
+        finally:
+            # if the there's a failure after yielding, don't fail to
+            # switch back to the prior alias but still raise whatever
+            # error condition occurred through the context
+            self._switch(prior, raise_not_found=False)
+
+    def _switch(self, alias: Optional[str], raise_not_found=True) -> Model:
+        if alias is None:
+            self._current_alias = None
```

Comment:
Is there a use case where setting alias=None is helpful?  If yes then this makes sense, but if no then this might just hide errors in the code that calls it (some unset alias value gets passed here)

Reply:
there may not be a reason for it.  Let me go back to the unit tests and ensure i can remove it

Reply:
I think this is resolved as "do not change" by the other `alias=None` discussion